### PR TITLE
RavenDB-12771 fixing resolved document and revisions behaviour

### PR DESCRIPTION
**Pull request by karmeli87**
- modifing a resovled document should keep the HasRevisions flags
- A revision that has the 'Conflicted' flag shouldn't overwrite the document
- fix various database change vector merging
- adding counter to a resovled document should not generate a revision

### CODE REVIEW

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 1]
```diff
@@ -250,9 +250,11 @@ public RevisionsCollectionConfiguration GetRevisionsConfiguration(string collect
             return Configuration.Default ?? _emptyConfiguration;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (ayende) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: This is a big method and not likely to be inlined even with this.
Why is this here?

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs [hunk 2]
```diff
@@ -312,6 +321,7 @@ public RevisionsCollectionConfiguration GetRevisionsConfiguration(string collect
             return true;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (ayende) [inline src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: remove this

**Review by ayende — COMMENTED** [review]
Talk to me about this tomorrow, I want to go over this with you so I can be sure that I understand what you have done here

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: done

File: src/Raven.Server/Documents/Revisions/RevisionsStorage.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Revisions/RevisionsStorage.cs]: done

File: src/Raven.Server/Documents/CountersStorage.cs [hunk 3]
```diff
@@ -693,7 +693,7 @@ public long GetNumberOfCounterEntries(DocumentsOperationContext context)
             if (hadModifications == false)
                 return;
 
-            var flags = doc.Flags.Strip(DocumentFlags.FromClusterTransaction);
+            var flags = doc.Flags.Strip(DocumentFlags.FromClusterTransaction | DocumentFlags.Resolved);
```
Comment (ayende) [inline src/Raven.Server/Documents/CountersStorage.cs:696]: Check what happens with attachments in the same scenario

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 4]
```diff
@@ -229,6 +222,20 @@ oldFlags .Contain(DocumentFlags.HasAttachments))
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: Remove this

File: src/Raven.Server/Documents/CountersStorage.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/CountersStorage.cs:696]: done

File: src/Raven.Server/Documents/DocumentPutAction.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/DocumentPutAction.cs]: done
